pageserver/src/walreceiver.rs

```diff
@@ -1,170 +1,567 @@
+                        if let Err(e) = connection.shutdown().await {
+                            error!("Failed to shutdown wal receiver connection for timeline {id}: {e:#}");
+                        }
+                    }
+                }
+                updates = subscription.fetch_data() => {
+                    if let Some(safekeeper_timeline_updates) = updates.and_then(|mut updates| updates.remove(&id)) {
+                        consider_walreceiver_connection(id, safekeeper_timeline_updates, &walreceiver_task_completion_sender,
+                            &timeline, connection_attempt, &mut wal_receiver_connection, connect_timeout).await;
+                    } else {
+                        info!("Subscription source end was dropped, no more updates are possible, shutting down");
+                        break;
+                    }
+                },
+            }
+        }
 
-    // Look up the current WAL producer address
-    let wal_producer_connstr = get_wal_producer_connstr(tenant_id, timeline_id);
+        shutdown_walreceiver(subscription, wal_receiver_connection).await.with_context(|| format!("Failed to shutdown walreceiver for timeline {id}"))
+    }.instrument(info_span!("timeline_walreceiver", id = %id)));
 
-    // Make a connection to the WAL safekeeper, or directly to the primary PostgreSQL server,
-    // and start streaming WAL from it.
-    let res = walreceiver_main(conf, tenant_id, timeline_id, &wal_producer_connstr);
+    Ok(TimelineWalReceiverHandles {
+        broker_join_handle,
+        cancellation_sender,
+    })
+}
 
-    // TODO cleanup info messages
-    if let Err(e) = res {
-        info!("WAL streaming connection failed ({})", e);
-    } else {
-        info!(
-            "walreceiver disconnected tenant {}, timelineid {}",
-            tenant_id, timeline_id
-        );
+async fn shutdown_walreceiver(
+    subscription: SkTimelineSubscription,
+    wal_receiver_connection: Option<WalReceiverConnection>,
+) -> anyhow::Result<()> {
+    subscription
+        .cancel()
+        .await
+        .context("Failed to cancel timeline subscription in etcd")?;
+    if let Some(connection) = wal_receiver_connection {
+        connection
+            .shutdown()
+            .await
+            .context("Failed to shutdown wal receiver connection")?;
     }
+    Ok(())
+}
+
+async fn consider_walreceiver_connection(
+    id: ZTenantTimelineId,
+    safekeeper_timelines: HashMap<ZNodeId, SkTimelineInfo>,
+    walreceiver_task_completion_sender: &Arc<watch::Sender<anyhow::Result<()>>>,
+    timeline: &DatadirTimelineImpl,
+    connection_attempt: u32,
+    connection: &mut Option<WalReceiverConnection>,
+    connect_timeout: Duration,
+) {
+    if let Some((new_safekeeper_id, new_safekeeper_timeline)) = safekeeper_timelines
+        .into_iter()
+        .max_by_key(|(_, info)| info.flush_lsn)
+        .filter(|(_, info)| info.flush_lsn > Some(timeline.tline.get_disk_consistent_lsn()))
```

Comment:
To avoid idle walreceivers, there is also should be logic for shutdown: if safekeepers stopped posting data about timeline, it means timeline became inactive and walreceiver should be stopped. We need to take into accounts deletes from etcd (happening after lease expires, soon after absense of new pushes) and remember old state for that.

(or invent some simpler way)

Reply:
Addressed and added some tests on top.
Tests are somewhat clumsy, but the whole thing seem to select safekeepers normally now.